mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -1740,3 +1741,294 @@ Value mlir::rock::addPassThroughIndices(OpBuilder &b, Value transformed,
   }
   return ret;
 }
+
+enum DimType { Upper = 0, Lower = 1 };
+
+struct TransformAttrArgs {
+  rock::TransformType type;
+  std::pair<SmallVector<StringRef>, SmallVector<StringRef>> perservedNames;
```

Comment:
Why not:
```
SmallVector<StringRef> preservedUpperNames;
SmallVector<StringRef> preservedLowerNames;
...
```
and then the utilities that need to be generic over upper or lower dimensions just take `ArrayRef<>` or `SmallVectorImpl<>&`?

This isn't a must-change sort of thing, but this tuples-and-templates structure just feels off

Reply:
```cpp
SmallVector<StringRef> preservedUpperNames;
SmallVector<StringRef> preservedLowerNames;
...
```
The above approach was my initial solution but it was too verbose. `std::pair` results in a more compact code which is  easy to read and, thus, to catch errors.  While working on this PR, I noticed that I made silly, typo-like mistakes quite often - e.g., using `preservedUpperNames` instead of `preservedLowerNames` and/or vice versa. The templated implementation is safer from this kind of errors. 

Conceptually, there is no difference. It is just a matter of taste but I would prefer the templated implementation because it is safer and more compact.